include/ringmesh/mesh/cartesian_grid.h

```diff
@@ -0,0 +1,526 @@
+            }
+            return ReferenceFrameManipulator<
+                DIMENSION >::coords_from_frame_to_global( cartesian_frame_,
+                cartesian_double_coords );
+        }
+
+        vecn< DIMENSION > local_coords_from_global_point(
+            const vecn< DIMENSION >& reference_vertex ) const
+        {
+            // Since coords_from_frame_to_global is faster than
+            // coords_from_global_to_frame, we use it with the inverse matrix
+            return ReferenceFrameManipulator< DIMENSION >::
+                coords_from_frame_to_global(
+                    inverse_cartesian_frame_, reference_vertex );
+        }
+
+        sivecn< DIMENSION > containing_cell_from_global_point(
+            const vecn< DIMENSION >& reference_vertex ) const
+        {
+            // Since coords_from_frame_to_global is faster than
+            // coords_from_global_to_frame, we use it with the inverse matrix
+            return this->containing_cell_from_local_point(
+                ReferenceFrameManipulator< DIMENSION >::
+                    coords_from_frame_to_global(
+                        inverse_cartesian_frame_, reference_vertex ) );
+        }
+
+        sivecn< DIMENSION > containing_cell_from_local_point(
+            const vecn< DIMENSION >& vertex ) const
+        {
+            sivecn< DIMENSION > coord;
+            for( auto i : range( DIMENSION ) )
+            {
+                coord[i] =
+                    static_cast< signed_index_t >( std::floor( vertex[i] ) );
+            }
+            return coord;
+        }
+
+        index_t cell_offset( const sivecn< DIMENSION >& coords ) const
+        {
+            index_t offset{ 0 };
+            index_t mult{ 1 };
+            for( auto i : range( DIMENSION ) )
+            {
+                if( coords[i] >= 0
+                    && coords[i] < static_cast< signed_index_t >(
+                                       nb_cells_in_each_direction_[i] ) )
+                {
+                    offset += static_cast< index_t >( coords[i] ) * mult;
+                    mult *= nb_cells_in_each_direction_[i];
+                }
+                else
+                {
+                    Logger::warn( "Point ", coords,
+                        " has indexes outside of the cartesian grid limits." );
+                    return NO_ID;
+                }
+            }
+            return offset;
+        }
+
+        index_t cell_offset_from_global_point(
+            const vecn< DIMENSION > coords ) const
+        {
+            return cell_offset( containing_cell_from_global_point( coords ) );
+        }
+
+        sivecn< DIMENSION > local_from_offset( const index_t offset ) const
+        {
+            sivecn< DIMENSION > coords;
+            index_t off{ 0 };
+            index_t div{ nb_total_cells_ };
+            for( auto i : range( DIMENSION ) )
+            {
+                div /= nb_cells_in_each_direction_[DIMENSION - 1 - i];
```

Comment:
Please add a check to assert that nb_cells_in_each_direction_[DIMENSION - 1 - i] is not equal to 0

Reply:
Moreover, why this decrement?

Reply:
It is already verified when creating the cartesian grid and whenever the user changes the number of cells in any direction.
The decrement is to be able to calculate the coordinates from the offset.